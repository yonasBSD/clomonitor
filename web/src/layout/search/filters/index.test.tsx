import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { vi } from 'vitest';

import Filters from './index';

const mockOnChange = vi.fn();
const mockOnChecksChange = vi.fn();
const mockOnAcceptedDateRangeChange = vi.fn();

const defaultProps = {
  visibleTitle: true,
  activeFilters: {},
  onChange: mockOnChange,
  onChecksChange: mockOnChecksChange,
  onAcceptedDateRangeChange: mockOnAcceptedDateRangeChange,
  device: 'test',
};

const realDate = Date;

const freezeDate = (isoDate: string) => {
  const fixedDate = new realDate(isoDate);

  const MockDate = class extends realDate {
    constructor(...args: ConstructorParameters<typeof realDate>) {
      if (args.length === 0) {
        super(fixedDate.getTime());
        return;
      }
      super(...args);
    }

    static now(): number {
      return fixedDate.getTime();
    }

    static parse(dateString: string): number {
      return realDate.parse(dateString);
    }

    static UTC(...args: Parameters<typeof realDate.UTC>): number {
      return realDate.UTC(...args);
    }
  };

  Object.setPrototypeOf(MockDate, realDate);
  // @ts-expect-error overriding global Date for tests
  global.Date = MockDate as unknown as DateConstructor;
};

describe('Filters', () => {
  beforeEach(() => {
    freezeDate('2022-03-24T00:00:00.000Z');
  });

  afterEach(() => {
    // @ts-expect-error restoring original Date
    global.Date = realDate;
    vi.resetAllMocks();
  });

  it('creates snapshot', () => {
    const { asFragment } = render(<Filters {...defaultProps} />);

    expect(asFragment()).toMatchSnapshot();
  });

  describe('Render', () => {
    it('renders Filters', () => {
      render(<Filters {...defaultProps} />);

      expect(screen.getByText('Filters')).toBeInTheDocument();

      expect(screen.getByText('Foundation')).toBeInTheDocument();
      expect(screen.getByRole('checkbox', { name: 'CNCF' })).toBeInTheDocument();
      expect(screen.getByRole('checkbox', { name: 'LF AI & Data' })).toBeInTheDocument();

      expect(screen.getByText('Rating')).toBeInTheDocument();
      expect(screen.getByRole('checkbox', { name: /A\s?\[75-100]/i })).toBeInTheDocument();
      expect(screen.getByRole('checkbox', { name: /B\s?\[50-74]/i })).toBeInTheDocument();
      expect(screen.getByRole('checkbox', { name: /C\s?\[25-49]/i })).toBeInTheDocument();
      expect(screen.getByRole('checkbox', { name: /D\s?\[0-24]/i })).toBeInTheDocument();

      expect(screen.getAllByText('From:')).toHaveLength(2);
      expect(screen.getAllByText('To:')).toHaveLength(2);
      expect(screen.getAllByText('Jan 1, 2016')).toHaveLength(2);
      expect(screen.getAllByText('Mar 24, 2022')).toHaveLength(2);

      expect(screen.getByText('Checks')).toBeInTheDocument();
      expect(screen.getByRole('button', { name: 'Open checks modal' })).toHaveTextContent('Add checks filters');
    });

    it('renders Filters', () => {
      render(<Filters {...defaultProps} activeFilters={{ foundation: ['cncf'] }} />);

      expect(screen.getByText('Filters')).toBeInTheDocument();

      expect(screen.getByText('Foundation')).toBeInTheDocument();
      expect(screen.getByRole('checkbox', { name: 'CNCF' })).toBeChecked();

      expect(screen.getByText('Maturity level')).toBeInTheDocument();
      expect(screen.getByRole('checkbox', { name: 'Graduated' })).toBeInTheDocument();
      expect(screen.getByRole('checkbox', { name: 'Incubating' })).toBeInTheDocument();
      expect(screen.getByRole('checkbox', { name: 'Sandbox' })).toBeInTheDocument();
    });

    it('renders Filters with selected options', () => {
      render(<Filters {...defaultProps} activeFilters={{ foundation: ['cncf'], rating: ['a', 'b'] }} />);

      expect(screen.getByRole('checkbox', { name: 'CNCF' })).toBeChecked();
      expect(screen.getByRole('checkbox', { name: /A\s?\[75-100]/i })).toBeChecked();
      expect(screen.getByRole('checkbox', { name: /B\s?\[50-74]/i })).toBeChecked();
    });

    it('calls onChange to click filter', async () => {
      render(<Filters {...defaultProps} />);

      const check = screen.getByRole('checkbox', { name: /A\s?\[75-100]/i });

      expect(check).not.toBeChecked();

      await userEvent.click(check);

      expect(mockOnChange).toHaveBeenCalledTimes(1);
      expect(mockOnChange).toHaveBeenCalledWith('rating', 'a', true);
    });
  });
});
