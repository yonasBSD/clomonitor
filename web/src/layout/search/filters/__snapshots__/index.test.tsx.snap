// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Filters > creates snapshot 1`] = `
<DocumentFragment>
  <div
    class="d-flex flex-row align-items-center justify-content-between pb-2 mb-4 border-bottom border-1"
  >
    <div
      class="h6 text-uppercase mb-0 lh-base text-primary fw-bold"
    >
      Filters
    </div>
  </div>
  <div
    class="fw-bold text-uppercase text-primary _categoryTitle_bcd004"
  >
    <small>
      Foundation
    </small>
  </div>
  <div
    class="mt-2"
  >
    <div
      class="form-check me-sm-2 mb-2 undefined"
    >
      <input
        aria-checked="false"
        class="form-check-input rounded-0 Checkbox-module_checkbox__lSeQj"
        id="test-foundation-cdf"
        name="foundation"
        tabindex="0"
        type="checkbox"
        value="cdf"
      />
      <label
        class="form-check-label mw-100"
        data-testid="checkboxLabel"
        for="test-foundation-cdf"
      >
        <div
          class="d-flex align-items-baseline mw-100"
        >
          <span
            class="position-relative _decorator_bcd004"
          />
          <span
            class="text-truncate"
          >
            CDF
          </span>
        </div>
      </label>
    </div>
    <div
      class="form-check me-sm-2 mb-2 undefined"
    >
      <input
        aria-checked="false"
        class="form-check-input rounded-0 Checkbox-module_checkbox__lSeQj"
        id="test-foundation-cncf"
        name="foundation"
        tabindex="0"
        type="checkbox"
        value="cncf"
      />
      <label
        class="form-check-label mw-100"
        data-testid="checkboxLabel"
        for="test-foundation-cncf"
      >
        <div
          class="d-flex align-items-baseline mw-100"
        >
          <span
            class="position-relative _decorator_bcd004"
          />
          <span
            class="text-truncate"
          >
            CNCF
          </span>
        </div>
      </label>
    </div>
    <div
      class="form-check me-sm-2 mb-2 undefined"
    >
      <input
        aria-checked="false"
        class="form-check-input rounded-0 Checkbox-module_checkbox__lSeQj"
        id="test-foundation-lfaidata"
        name="foundation"
        tabindex="0"
        type="checkbox"
        value="lfaidata"
      />
      <label
        class="form-check-label mw-100"
        data-testid="checkboxLabel"
        for="test-foundation-lfaidata"
      >
        <div
          class="d-flex align-items-baseline mw-100"
        >
          <span
            class="position-relative _decorator_bcd004"
          />
          <span
            class="text-truncate"
          >
            LF AI & Data
          </span>
        </div>
      </label>
    </div>
    <div
      class="form-check me-sm-2 mb-2 undefined"
    >
      <input
        aria-checked="false"
        class="form-check-input rounded-0 Checkbox-module_checkbox__lSeQj"
        id="test-foundation-lfnetworking"
        name="foundation"
        tabindex="0"
        type="checkbox"
        value="lfnetworking"
      />
      <label
        class="form-check-label mw-100"
        data-testid="checkboxLabel"
        for="test-foundation-lfnetworking"
      >
        <div
          class="d-flex align-items-baseline mw-100"
        >
          <span
            class="position-relative _decorator_bcd004"
          />
          <span
            class="text-truncate"
          >
            LF Networking
          </span>
        </div>
      </label>
    </div>
  </div>
  <div
    class="fw-bold text-uppercase text-primary _categoryTitle_bcd004"
  >
    <small>
      Rating
    </small>
  </div>
  <div
    class="mt-2"
  >
    <div
      class="form-check me-sm-2 mb-2 undefined"
    >
      <input
        aria-checked="false"
        class="form-check-input rounded-0 Checkbox-module_checkbox__lSeQj"
        id="test-rating-a"
        name="rating"
        tabindex="0"
        type="checkbox"
        value="a"
      />
      <label
        class="form-check-label mw-100"
        data-testid="checkboxLabel"
        for="test-rating-a"
      >
        <div
          class="d-flex align-items-baseline mw-100"
        >
          <span
            class="position-relative _decorator_bcd004"
          >
            <div
              class="me-2 border border-1 rounded-pill position-relative _quality_d52a11 _level1_d52a11 undefined"
              data-testid="quality-dot"
            />
          </span>
          <span
            class="text-truncate"
          >
            A
          </span>
          <small
            class="ps-1"
          >
            [75-100]
          </small>
        </div>
      </label>
    </div>
    <div
      class="form-check me-sm-2 mb-2 undefined"
    >
      <input
        aria-checked="false"
        class="form-check-input rounded-0 Checkbox-module_checkbox__lSeQj"
        id="test-rating-b"
        name="rating"
        tabindex="0"
        type="checkbox"
        value="b"
      />
      <label
        class="form-check-label mw-100"
        data-testid="checkboxLabel"
        for="test-rating-b"
      >
        <div
          class="d-flex align-items-baseline mw-100"
        >
          <span
            class="position-relative _decorator_bcd004"
          >
            <div
              class="me-2 border border-1 rounded-pill position-relative _quality_d52a11 _level2_d52a11 undefined"
              data-testid="quality-dot"
            />
          </span>
          <span
            class="text-truncate"
          >
            B
          </span>
          <small
            class="ps-1"
          >
            [50-74]
          </small>
        </div>
      </label>
    </div>
    <div
      class="form-check me-sm-2 mb-2 undefined"
    >
      <input
        aria-checked="false"
        class="form-check-input rounded-0 Checkbox-module_checkbox__lSeQj"
        id="test-rating-c"
        name="rating"
        tabindex="0"
        type="checkbox"
        value="c"
      />
      <label
        class="form-check-label mw-100"
        data-testid="checkboxLabel"
        for="test-rating-c"
      >
        <div
          class="d-flex align-items-baseline mw-100"
        >
          <span
            class="position-relative _decorator_bcd004"
          >
            <div
              class="me-2 border border-1 rounded-pill position-relative _quality_d52a11 _level3_d52a11 undefined"
              data-testid="quality-dot"
            />
          </span>
          <span
            class="text-truncate"
          >
            C
          </span>
          <small
            class="ps-1"
          >
            [25-49]
          </small>
        </div>
      </label>
    </div>
    <div
      class="form-check me-sm-2 mb-2 undefined"
    >
      <input
        aria-checked="false"
        class="form-check-input rounded-0 Checkbox-module_checkbox__lSeQj"
        id="test-rating-d"
        name="rating"
        tabindex="0"
        type="checkbox"
        value="d"
      />
      <label
        class="form-check-label mw-100"
        data-testid="checkboxLabel"
        for="test-rating-d"
      >
        <div
          class="d-flex align-items-baseline mw-100"
        >
          <span
            class="position-relative _decorator_bcd004"
          >
            <div
              class="me-2 border border-1 rounded-pill position-relative _quality_d52a11 _level4_d52a11 undefined"
              data-testid="quality-dot"
            />
          </span>
          <span
            class="text-truncate"
          >
            D
          </span>
          <small
            class="ps-1"
          >
            [0-24]
          </small>
        </div>
      </label>
    </div>
  </div>
  <div>
    <div
      class="d-flex flex-row align-items-center text-primary _categoryTitle_512c5a"
    >
      <div
        class="fw-bold text-uppercase"
      >
        <small>
          Checks
        </small>
      </div>
    </div>
    <div>
      <button
        aria-label="Open checks modal"
        class="btn btn-outline-secondary btn-sm rounded-0 _btn_512c5a mt-3"
        type="button"
      >
        <span>
          Add checks filters
        </span>
      </button>
    </div>
  </div>
  <div>
    <div>
      <label>
        From:
      </label>
      <span>
        Jan 1, 2016
      </span>
    </div>
    <div>
      <label>
        To:
      </label>
      <span>
        Mar 24, 2022
      </span>
    </div>
    <div>
      <label>
        From:
      </label>
      <span>
        Jan 1, 2016
      </span>
    </div>
    <div>
      <label>
        To:
      </label>
      <span>
        Mar 24, 2022
      </span>
    </div>
  </div>
</DocumentFragment>
`;
